### PR TITLE
[AIRFLOW-6497] Avoid loading DAGs in the main scheduler loop

### PR DESCRIPTION
**Pull request by mik-laj**
This change should speed up the main scheduler loop, because it will not load the DAG file. This should also drastically reduce the amount of memory used because modules with external operators and libraries will not be loaded.

Thanks for support to @evgenyshulman from Databand!

---
Issue link: [AIRFLOW-6497](https://issues.apache.org/jira/browse/AIRFLOW-6497)

Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: tests/dags/test_on_failure_callback.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    'owner': 'airflow',
+    'start_date': DEFAULT_DATE,
+}
+
+dag = DAG(dag_id='test_om_failure_callback_dag', default_args=args)
+
+
+def write_data_to_callback(*arg, **kwargs):  # pylint: disable=unused-argument
+    with open(os.environ.get('AIRFLOW_CALLBACK_FILE'), "w+") as f:
+        f.write("Callbacck fired")
```
Comment (tooptoop4) [inline tests/dags/test_on_failure_callback.py]: cck typo

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -1015,14 +1012,52 @@ def test_kill_zombies(self, mock_ti_handle_failure):
...
+                    msg="Message"
+                )
+            ]
+            callback_file.close()
+
+            with mock.patch.dict("os.environ", {"AIRFLOW_CALLBACK_FILE": callback_file.name}):
+                dag_file_processor.process_file(dag_file, requests)
+            with open(callback_file.name) as callback_file2:
+                content = callback_file2.read()
+            self.assertEqual("Callbacck fired", content)
```
Comment (tooptoop4) [inline tests/jobs/test_scheduler_job.py]: cck typo

**Review by tooptoop4 — CHANGES_REQUESTED** [review]

File: tests/jobs/test_scheduler_job.py (hunk 2 above)
Comment (mik-laj) [reply tests/jobs/test_scheduler_job.py]: Done

**Comment by ashb** [comment]
> This change should speed up the main scheduler loop

Do you have any numbers for this please?

File: airflow/jobs/scheduler_job.py [hunk 3]
```diff
@@ -1616,26 +1633,6 @@ def _validate_and_run_task_instances(self, simple_dag_bag: SimpleDagBag) -> bool
         self._process_executor_events(simple_dag_bag)
         return True
 
-    def _process_and_execute_tasks(self, simple_dag_bag):
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py]: Moving code around like this makes PRs harder to review -- as you are mixing two different things here and making more work for reviewers -- and this change is unrelated to "not loading DAGs in main scheduler loop".

File: airflow/utils/dag_processing.py [hunk 4]
```diff
@@ -655,6 +693,7 @@ def start(self):
             # Update number of loop iteration.
             self._num_run += 1
 
+            simple_dags = self.collect_results()
```
Comment (ashb) [inline airflow/utils/dag_processing.py:696]: Was there a reason you moved the position of this call?

Comment (mik-laj) [reply airflow/utils/dag_processing.py:696]: Yes. First we need to create processes, and then we can read the results. Otherwise, we will never get the value on the first iteration of the loop. Other solution: we can increase loop iteration nunber: https://github.com/apache/airflow/blob/cb455dc81162680f90edcd78400e1ef46c09766d/tests/utils/test_dag_processing.py#L291

**Comment by mik-laj** [comment]
> Do you have any numbers for this please?

It is very difficult to measure because it depends on the specific DAG File. Some DAG files take up to 30 seconds or more to load. During this time, the scheduler loop is stopped and does not start any new tasks.  I can measure how long it takes to load example_dags, but it's not just a subset of cases. It doesn't provide real values,... but I created a spreadsheet:
When I ran the following script:
```python
import os
import sys
import time
from contextlib import contextmanager

import psutil

from airflow.models import DagBag


@contextmanager
def timing_ctx():
    time1 = time.time()
    try:
        yield
    finally:
        time2 = time.time()
        diff = (time2 - time1) * 1000.0
        print('Time: %0.3f ms' % diff)


def get_process_memory():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss


@contextmanager
def memory_ctx():
    before = get_process_memory()
    try:
        yield
    finally:
        after = get_process_memory()
        diff = after - before
        print('Memory: %d bytes' % diff)


filename = sys.argv[1]

with timing_ctx(), memory_ctx():
    print("Filename:", filename)
    DagBag(dag_folder=filename, include_examples=False, store_serialized_dags=False)
```
```
find  airflow/providers/google/cloud/example_dags/ -type f | sort| grep -v "__init__.py" | grep -v "__init__.py" | xargs -n 1 readlink -e  | xargs -t -n 1 python /files/performance/load_dag_perf_test.py
```
I got following values:
https://docs.google.com/spreadsheets/d/1T0kLEQLSU5ujxU-W_PoxddbkEgWx70EQkpwjRLNWaic/edit?usp=sharing

**Comment by mik-laj** [comment]
It is worth noting that this also solves one more problem. The modules are always reloaded.
https://issues.apache.org/jira/projects/AIRFLOW/issues/AIRFLOW-6497?filter=allopenissues
so when someone makes a change in the additional module it is correctly executed. Its old version is not executed. This can be a problem because the handler is often stored in helper functions and is shared among many DAGs.

File: airflow/jobs/scheduler_job.py (hunk 3 above)
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: I will revert this change and propose it in a separate PR.

**Comment by mik-laj** [comment]
This also protects against killing the scheduler by incorrect error handling functions.  If sys.exit(1) appears in the handler code, the scheduler will not be stopped.  It will only affect one DAG.

File: airflow/utils/dag_processing.py (hunk 4 above)
Comment (ashb) [reply airflow/utils/dag_processing.py:696]: Cool, sounds good as you have it

File: airflow/jobs/scheduler_job.py [hunk 5]
```diff
@@ -752,37 +758,35 @@ def _find_dags_to_process(self, dags: List[DAG], paused_dag_ids: Set[str]) -> Li
...
-                    ti.state = zombie.state
+                    ti.start_date = request.simple_task_instance.start_date
+                    ti.end_date = request.simple_task_instance.end_date
+                    ti.try_number = request.simple_task_instance.try_number
+                    ti.state = request.simple_task_instance.state
                     ti.test_mode = self.UNIT_TEST_MODE
-                    ti.handle_failure("{} detected as zombie".format(ti),
-                                      ti.test_mode, ti.get_template_context())
-                    self.log.info('Marked zombie job %s as %s', ti, ti.state)
-                    Stats.incr('zombies_killed')
```
Comment (kaxil) [inline airflow/jobs/scheduler_job.py]: Needs mention in API and we need to remove `zombies_killed` from the metrics list

File: airflow/jobs/scheduler_job.py [hunk 6]
```diff
@@ -752,37 +758,35 @@ def _find_dags_to_process(self, dags: List[DAG], paused_dag_ids: Set[str]) -> Li
...
         TI = models.TaskInstance
 
-        for zombie in zombies:
-            if zombie.dag_id in dagbag.dags:
-                dag = dagbag.dags[zombie.dag_id]
-                if zombie.task_id in dag.task_ids:
-                    task = dag.get_task(zombie.task_id)
-                    ti = TI(task, zombie.execution_date)
+        for request in failure_callback_requests:
+            if request.simple_task_instance.dag_id in dagbag.dags:
```
Comment (kaxil) [inline airflow/jobs/scheduler_job.py]: We can create a local var `simple_ti` to avoid repeating `request.simple_task_instance` in this loop

Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: Done

**Review by tooptoop4 — APPROVED** [review]

**Review by KevinYang21 — APPROVED** [review]
I love this change! 👍 👍 👍 I can add one data point: we've seen multiple times in our cluster that when there's spike on executor events, scheduler loop will take forever, e.g. we have 300 executor event from a heavy DAG, it takes 30 seconds to parse the DAG, it would take 2h+ to finish handling all the executor events. Of c in reality we never waited so long but restart the scheduler after 30 mins when scheduler loop stuck here.

We attempted similar thing before but never followed it through. The concern we had previously was that delegate the on failure execution can change the existing behavior. Previously it was a blocking call in the main loop where task failure will be handled before we start the next scheduling loop and now the failure handling can be delayed( was an issue for us because we have a patch to aggresssively reschedule tasks). I don't think it's an issue in the master tho.

**Review by BasPH — APPROVED** [review]

**Review by kaxil — COMMENTED** [review]
CI is failing. Once it passes, it can be merged

**Comment by mik-laj** [comment]
Travis is green.

**Comment by ashb** [comment]
@mik-laj Am I following the path of `FailureCallbackRequest` correctly for zombie requests? Does it start in the DagFileProcessorManager, get sent back via the Agent to the scheduler, only to the get sent _back_ to the ProcessorManager?

**Comment by mik-laj** [comment]
1. For executor events

https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/jobs/scheduler_job.py#L1430
Send to DagFileProcessorManager via DagFileProcesorAgent
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/jobs/scheduler_job.py#L1464-L1469
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/utils/dag_processing.py#L670-L673
Send to queue
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/utils/dag_processing.py#L737-L742
Start a new DagFileProcessor via DagFileProcessorProcess
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/utils/dag_processing.py#L1045-L1057
Execute callback
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/jobs/scheduler_job.py#L835-L838

2. For zombies

Find zombies
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/utils/dag_processing.py#L670-L673
Send to queue
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/utils/dag_processing.py#L1139
Start a new DagFileProcessor via DagFileProcessorProcess
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/utils/dag_processing.py#L1045-L1057
Execute callback
https://github.com/PolideaInternal/airflow/blob/99d237b684bbcdcf808451dfbfad425c0c7f772b/airflow/jobs/scheduler_job.py#L835-L838

**Comment by mik-laj** [comment]
For processor events
Scheduler send the callback to DagFileProcessorManager, DagFileProcessorManager starts DagFileProcessor via DagFileProcessorProcess and process callback
Scheduler => DagFileProcessorManager => DagFileProcessor => DagFileProcessorProcess
For zombies:
DagFileProcessorManager => DagFileProcessor => DagFileProcessorProcess

**Comment by ashb** [comment]
Oh sorry, I misread this section of tests https://github.com/apache/airflow/blob/fc862a3edd010e65b9b3fe586855fe81807ee4e8/tests/utils/test_dag_processing.py#L225-L229

I thought that was expecting the failure callback to also come back, but that's not what it's checking

**Comment by feng-tao** [comment]
just learn this pr, nice!
